source/ncengine/physics/jolt/VehicleManager.cpp

```diff
@@ -0,0 +1,155 @@
+#include "VehicleManager.h"
+#include "ConstraintFactory.h"
+
+#include "Jolt/Jolt.h"
+#include "Jolt/Physics/PhysicsSystem.h"
+#include "Jolt/Physics/Vehicle/WheeledVehicleController.h"
+#include "Jolt/Physics/Vehicle/VehicleConstraint.h"
+
+#include <ranges>
+
+namespace
+{
+void AddToSimulation(JPH::PhysicsSystem* physicsSystem,
+                     JPH::VehicleConstraint* constraint)
+{
+    physicsSystem->AddConstraint(constraint);
+    physicsSystem->AddStepListener(constraint);
+}
+
+void RemoveFromSimulation(JPH::PhysicsSystem* physicsSystem,
+                          JPH::VehicleConstraint* constraint)
+{
+    physicsSystem->RemoveStepListener(constraint);
+    physicsSystem->RemoveConstraint(constraint);
+}
+} // anonymous namesapce
+
+namespace nc::physics
+{
+VehicleManager::VehicleManager(JPH::PhysicsSystem& physicsSystem,
+                               ConstraintFactory& constraintFactory,
+                               uint32_t maxEntities)
+    : m_physicsSystem{&physicsSystem},
+      m_factory{&constraintFactory},
+      m_entityToVehicle{std::min(ConstraintMapSizeHint, maxEntities), maxEntities}
+{
+}
+
+VehicleManager::~VehicleManager() noexcept = default;
+
+auto VehicleManager::AddVehicle(VehicleInfo&& createInfo,
+                                Entity owner,
+                                JPH::Body& ownerBody) -> Vehicle&
+{
+    NC_ASSERT(!GetVehicle(owner), "RigidBody already has a Vehicle");
+    auto vehicle = m_factory->MakeVehicleConstraint(createInfo, ownerBody); // careful, this gets moved
+    auto vehiclePtr = vehicle.GetPtr();
+    if (!m_isBatchInProgress)
+    {
+        AddToSimulation(m_physicsSystem, vehiclePtr);
+    }
+
+    const auto index = [&]()
+    {
+        if (m_freeIndices.empty() || m_isBatchInProgress)
+        {
+            const auto i = static_cast<uint32_t>(m_handles.size());
+            m_handles.push_back(std::move(vehicle));
+            return i;
+        }
+
+        const auto i = m_freeIndices.back();
+        m_freeIndices.pop_back();
+        m_handles[i] = std::move(vehicle);
+        return i;
+    }();
+
+    const auto ownerId = owner.Index();
+    auto& state = m_entityToVehicle.contains(ownerId)
+        ? m_entityToVehicle.at(ownerId)
+        : m_entityToVehicle.emplace(ownerId, {});
+
+    state = std::make_unique<Vehicle>(std::move(createInfo), vehiclePtr, index);
```

Comment:
I see your comment about vehicle being moved and the std::move - how is this pointer valid at this point? Can you explain?

Reply:
Its a ref counted pointer, so its just like:
```cpp
auto smart = std::make_unique<Foo>();
auto raw = smart.get();
Consume(std::move(smart));
raw->Use(); // ok
smart->Use(); // nullptr dereference
```
where `move` changes the owner, but the address on the heap is the same.